cylc/flow/scripts/lint.py

```diff
@@ -0,0 +1,514 @@
+in files.
+
+When run with "-i" ("--inplace") mode this will alter your files.
+We strongly recommend committing your workflow to version control
+before using "cylc lint -i".
+
+Examples:
+
+# run as a linter
+cylc lint <paths to workflow directories to check>
+
+# run inplace
+cylc lint --inplace <paths to workflow directories to check>
+cylc lint -i <paths to workflow directories to check>
+
+# Get information about errors:
+cylc lint --reference
+cylc lint -R
+"""
+from colorama import Fore
+from optparse import Values
+from pathlib import Path
+import re
+from typing import Generator
+
+from cylc.flow import LOG
+from cylc.flow.option_parsers import (
+    CylcOptionParser as COP,
+)
+from cylc.flow.cfgspec.workflow import upg, SPEC
+from cylc.flow.parsec.config import ParsecConfig
+from cylc.flow.terminal import cli_function
+
+STYLE_GUIDE = (
+    'https://cylc.github.io/cylc-doc/latest/html/workflow-design-guide/'
+    'style-guide.html#'
+)
+URL_STUB = "https://cylc.github.io/cylc-doc/latest/html/7-to-8/"
+SECTION1 = r'\[\s*{}\s*\]'
+SECTION2 = r'\[\[\s*{}\s*\]\]'
+SECTION3 = r'\[\[\[\s*{}\s*\]\]\]'
+FILEGLOBS = ['*.rc', '*.cylc']
+JINJA2_SHEBANG = '#!jinja2'
+DEPENDENCY_SECTION_MSG = {
+    'text': (
+        '"[scheduling][dependencies][<recurrence>]graph =" -> '
+        '"[scheduling][graph]<recurrence> ="'
+    ),
+    'rst': (
+        '``[scheduling][dependencies][<recurrence>]graph =`` -> '
+        '``[scheduling][graph]<recurrence> =``'
+    )
+}
+JOBANDREMOTE_SECTION_MSG = {
+    'text': (
+        'settings in "[runtime][<namespace>][{}]" have been moved to '
+        '"[runtime][<namespace>]" and "global.cylc[platforms]'
+        '[<platforms name>]"'
+    ),
+    'rst': (
+        'settings in ``[runtime][<namespace>][{}]`` have been moved to '
+        '``[runtime][<namespace>]`` and ``global.cylc[platforms]'
+        '[<platforms name>]``'
+    )
+}
+JINJA2_FOUND_WITHOUT_SHEBANG = 'jinja2 found: no shebang (#!jinja2)'
+CHECKS_DESC = {'U': '7 to 8 upgrades', 'S': 'Style'}
+STYLE_CHECKS = {
+    re.compile(r'^\t'): {
+        'short': 'Use multiple spaces, not tabs',
+        'url': STYLE_GUIDE + 'tab-characters'
+    },
+    # Not a full test, but if a non section is not indented...
+    re.compile(r'^[^\{\[|\s]'): {
+        'short': 'Item not indented.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
+    #            [section]
+    re.compile(r'^\s+\[[^\[.]*\]'): {
+        'short': 'Too many indents for top level section.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
+    # 2 or 4 space indentation both seem reasonable:
+    re.compile(r'^(\s|\s{3}|\s{5,})\[\[[^\[.]*\]\]'): {
+        'short': 'wrong number of indents for second level section.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
+    re.compile(r'^(\s{1,3}|\s{5,7}|\s{9,})\[\[\[[^\[.]*\]\]\]'): {
+        'short': 'wrong number of indents for third level section.',
+        'url': STYLE_GUIDE + 'indentation'
+    },
```

Comment:
Did we agree to keep the indentation checks? I thought it would be very noisy due to people indenting in between Jinja2 expressions

Reply:
@oliver-sanders  what do you think?

Reply:
I've made them fairly liberal, we could make them more relaxed still (odd numbers of indentations will always be wrong)

Reply:
From team meeting:
1. These checks are OK wrt jinja2 indentation - style guide says no.
2. Simplify these to only allow 4*X indents.

Reply:
(Although it won't ensure consistency with the settings that come under the section)